File: services/server/src/safe_socket/safe_socket.py
import socket

# TODO: Complete with a short-read/short-write tolerant implementation


def recv_all(sock: socket.socket, size):
    buffer = b""
    while len(buffer) < size:
        chunk = sock.recv(size - len(buffer))
        if not chunk:
            raise ConnectionError(
                f"Conexion cerrada: recibidos {len(buffer)}/{size} bytes"
            )
        buffer += chunk
    return buffer


def send_all(sock: socket.socket, data):
    total_sent = 0  #bytes enviados
    while total_sent < len(data):
        sent = sock.send(data[total_sent:])
        total_sent += sent
    return total_sent

File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import "io"

//TODO: Complete with a short-read/short-write tolerant implementation

func SendAll(socket io.Writer, bytes []byte) error {
	totalSent := 0
	for totalSent < len(bytes) {
		n, err := socket.Write(bytes[totalSent:])
		if err != nil {
			return err
		}
		// Nota: un Writer puede legítimamente devolver n == 0 sin error
		// (I/O no garantiza progreso en una sola llamada), por eso NO se
		// trata como un error fatal: simplemente se reintenta hasta
		// completar el envío.
		totalSent += n
	}
	return nil
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	buff := make([]byte, size)
	totalRead := 0
	for totalRead < size {
		n, err := socket.Read(buff[totalRead:])
		totalRead += n
		if totalRead == size {
			return buff[:totalRead], nil
		}
		if err != nil {
			return nil, err
		}
		if n == 0 {
			return nil, io.ErrUnexpectedEOF
		}
	}
	return buff[:totalRead], nil
}
